File: Data Structures & Algorithms/house-robber-iii/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map <TreeNode*, int> memo;
    int rob(TreeNode* root) {
        return dfs(root);
    }
    int dfs(TreeNode *root) {
        if (memo.count(root) == 1) return memo[root];
        if (!root) return 0;
        memo[root] = root->val;
        if (root->left != nullptr) {
            memo[root] += dfs(root->left->left) + dfs(root->left->right);
        }
        if (root->right) {
            memo[root] += dfs(root->right->left) + dfs(root->right->right);
        }
        memo[root] = max(memo[root], dfs(root->left) + dfs(root->right));
        return memo[root];
    }
};

File: Data Structures & Algorithms/two-sum-bsts/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def twoSumBSTs(self, root1: Optional[TreeNode], root2: Optional[TreeNode], target: int) -> bool:
        def dfs(p, q):
            if not p or not q:
                return False
            if p.val + q.val == target:
                return True
            if p.val + q.val > target:
                return dfs(p.left, q) or dfs(p, q.left)
            else:
                return dfs(p.right, q) or dfs(p, q.right)
        return dfs(root1, root2)